File: src/json.rs
use serde::{Serialize};

#[derive(Serialize)]
pub struct JsonResponse<'a, T: Serialize> {
    pub success: bool,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<&'a T>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub message: Option<String>,
}

pub fn json_ok<T: Serialize>(data: Option<&T>) -> JsonResponse<T> {
    JsonResponse {
        success: true,
        data,
        message: None
    }
}

pub fn json_error(message: &str) -> JsonResponse<String> {
    JsonResponse {
        success: false,
        data: None,
        message: Some(message.to_string())
    }
}

File: src/json.ts
export interface JsonOk<T> {
    success: true;
    data?: T;
}

export interface JsonErr<T> {
    success: false;
    message: string;
    additionalInformation?: T;
}

export function Ok<T = void>(data?: T): JsonOk<T> {
    if (typeof data === 'undefined') {
        return {success: true};
    } else {
        return {success: true, data};
    }
}

export function Err<T = void>(message: string, additionalInformation?: T): JsonErr<T> {
    if (typeof additionalInformation === 'undefined') {
        return {success: false, message};
    } else {
        return {success: false, message, additionalInformation};
    }
}
